### PR TITLE
Enable trimming TLS arrays in ArrayPool.Shared

### PR DESCRIPTION
**Pull request by stephentoub**
Today arrays stored in `ArrayPool<T>.Shared`'s per-core buckets have trimming applied, but arrays stored in the per-thread buckets are only trimmed when there's high memory pressure.  This change enables all buffers to be trimmed (eventually).  Every time our gen2 callback runs, it ensures any non-timestamped buffers have a timestamp, and also ensures that any timestamped buffers are still timely... if any aren't, they're eligible for trimming.  The timestamp is reset for TLS arrays when they're stored, and for per-core buckets when they transition from empty to non-empty; the latter is just a tweak on the current behavior, which incurs the cost of Environment.TickCount upon that transition, whereas now we only pay it as part of the trimming pass.

Contributes to https://github.com/dotnet/runtime/issues/52098
cc: @jkotas, @Maoni0

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:499]: Should this rather be `trimTicks / 4`? Otherwise, the `_timestamp` will be in future for `MemoryPressure.High` that will be treated as tickcount wrap around.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:357]:
```suggestion
            public void Trim(int milliseconds, int id, Utilities.MemoryPressure pressure, int bucketSize)
```
The rest of the file uses `milliseconds` for the current tick count variable. It may be nice to do it everywhere for consistency.

**Review by jkotas — APPROVED** [review]
Thank you!

File: src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:271]: Ok.  I'd done this to mimic what was already being done in Trim, but I'll remove it from there, too.

**Comment by vargaz** [comment]
The failures are caused by a linker problem. Here is the original finally clause in ::Trim ():
```
     finally
    {
      IL_009a:  ldloc.s    V_5
      IL_009c:  brfalse.s  IL_00a5

      IL_009e:  ldloc.s    V_5
      IL_00a0:  callvirt   instance void System.IDisposable::Dispose()
      IL_00a5:  endfinally
    }  // end handler
```
and here is the linked version:
```
    finally
    {
      IL_0091:  ldloc.s    V_5
      IL_0093:  brfalse.s  IL_009c

      IL_0095:  ldloc.s    V_5
      IL_0097:  callvirt   instance void System.IDisposable::Dispose()
      IL_009c:  endfinally
      IL_009d:  ldloc.1
      IL_009e:  ldc.i4.1
      IL_009f:  bne.un.s   IL_00aa

      IL_00a1:  ldc.i4     0x3a98
    }  // end handler
```

**Comment by stephentoub** [comment]
> The failures are caused by a linker problem

cc: @sbomer, @vitek-karas

**Comment by danmoseley** [comment]
@antonfirsov

File: src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs
Comment (antonfirsov) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:248]: More a question than anything else:

The `Trim()` code has grown relatively complex. Is it fine to run all this on the finalizer thread? Wouldn't it be better to delegate the actual work to ThreadPool? Or is it negligible considering the whole queue?

**Comment by stephentoub** [comment]
@steveisok, how can we unblock this PR?

File: src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:248]: It's not adding a lot of work.  It'd be an ok thing to explore separately, but queueing it to the thread pool would also enable concurrent invocations of Trim, and the implementation would need to be audited to make sure that worked correctly.  It would also mean that trimming could be delayed not just by lots of objects to be finalized but also by lots of queued work to the pool, which could have harder to understand emergent behavior.  And this only fired because we're doing a gen2 GC, which means we're already doing a lot of work as part of that.

**Comment by danmoseley** [comment]
@stephentoub failures in runtime-staging configurations do not block merge. Although, I thought they should not show up as red for that reason.

**Comment by steveisok** [comment]
> @stephentoub failures in runtime-staging configurations do not block merge. Although, I thought they should not show up as red for that reason.

That's true for test failures only.  If you have a build failure, then it'll block the PR.

**Comment by danmoseley** [comment]
(I mean, they don't block merge if the failures aren't clearly related. If this change breaks them more, we probably don't want to merge it.)

**Comment by danmoseley** [comment]
That makes sense.

**Comment by stephentoub** [comment]
> I mean, they don't block merge if the failures aren't clearly related. If this change breaks them more, we probably don't want to merge it.

Right.  The problem is that the correct C# code and generated IL introduced by this PR is then getting mutilated by the linker and causing the AOT compiler to seg fault.  As a temporary workaround to unblock this PR, I'd be happy to tweak the C# code to avoid whatever pattern is tripping up the linker, if someone can suggest what that pattern is.  But it seems to me there's both a bug in the linker (https://github.com/mono/linker/issues/2181) and a bug in the AOT compiler (seg faulting in response to bad IL) to be fixed here.

**Comment by jkotas** [comment]
> if someone can suggest what that pattern is.

My guess is that ifdefing the logging specific paths in Trim method for Mono is going to workaround the linker bug:

```
#if !MONO // Workaround for http://github.com/...
     if (log.IsEnabled())
     {
     ....
     }
     else
#endif

**Comment by stephentoub** [comment]
Ok, thanks, I'll give that a go.

**Comment by DrewScoggins** [comment]
We are seeing some wins in the ArrayPool perf tests as a result of this change.

image

**Comment by stephentoub** [comment]
Thanks, @DrewScoggins.  It wasn't the initial intention of the PR, but I'm glad it had this affect (I expect from removing the TickCount check every time we transitioned from 0 to 1 array stored in the per-core stack).  Do you see any movement here that might counteract what you reported in https://github.com/dotnet/runtime/issues/56555?

**Comment by DrewScoggins** [comment]
Yes, we saw some of it come back, but not all.

image

**Pull request merged**
